nomad/eval_broker.go

```diff
@@ -934,3 +969,43 @@ func (p PendingEvaluations) Peek() *structs.Evaluation {
 	}
 	return p[n-1]
 }
+
+// PopN removes and returns the minimum N evaluations from the slice.
+func (p *PendingEvaluations) PopN(n int) []*structs.Evaluation {
+	if n > len(*p) {
+		n = len(*p)
+	}
+
+	popped := []*structs.Evaluation{}
+	for i := 0; i < n; i++ {
+		raw := heap.Pop(p)
+		eval := raw.(*structs.Evaluation)
+		popped = append(popped, eval)
+	}
+	return popped
+}
+
+// MarkForCancel is used to remove any evaluations older than the index from the
+// blocked list and returns a list of cancelable evals so that Eval.Ack RPCs can
+// write batched raft entries to cancel them. This must be called inside the
+// broker's lock.
+func (p *PendingEvaluations) MarkForCancel(index uint64) PendingEvaluations {
+
+	// In pathological cases, we can have a large number of blocked evals but
+	// will want to cancel most of them. Using heap.Remove requires we re-sort
+	// for each eval we remove. Because we expect to have very few if any
+	// blocked remaining, we'll just create a new heap
+	retain := PendingEvaluations{}
+	cancelable := PendingEvaluations{}
+
+	for _, eval := range *p {
+		if eval.ModifyIndex >= index {
+			heap.Push(&retain, eval)
+		} else {
+			heap.Push(&cancelable, eval)
+		}
+	}
+
+	*p = retain
+	return cancelable
+}
```

Comment:
Hold on... due to my "favorite" logic in the `Eval.Dequeue` RPC: https://github.com/hashicorp/nomad/blob/v1.4.1/nomad/eval_endpoint.go#L185-L211

...I think we can cancel *every eval except the one with the highest ModifyIndex?!*

Since on Dequeue we always set `WaitIndex = max(ModifyIndex)` it doesn't matter which eval we dequeue next for a job: we'll always process it against a new enough snapshot to capture every cluster change that triggered evals!

...it almost makes me think we could cancel these evals in the FSM, but we'd still need to cancel them in EvalBroker since the broker isn't a "live" view of the evals in the statestore... might not be impossible, but when in doubt I like to keep the FSM simple.

Reply:
I think you're right on this, but I'm going to work up a diagram of what indexes we have and what eval states we have in the broker just to double-check. 

My next instinct here is to mark the eval as cancelable at the time we enqueue rather than at the time we ack. Have `EvalBroker.blocked` contain a spot for at most one eval, and have `EvalBroker.Enqueue` drop the eval into `EvalBroker.cancelable` immediately if it's full. (As you've said, I'd love to be able to have the FSM detect that and delete the eval right away as well, but the followers wouldn't see the eval broker state. This would be the next best thing.) Let me work up that diagram first and we can come back to this discussion.

Reply:
I think one of the things that's misleading about the existing eval broker code is the `PendingEvaluations` sorting by `CreateIndex`. The `PendingEvaluations` alias over a slice of evals gets used by both `blocked` and `ready` but only `ready` actually cares about the index b/c it contains evals from multiple jobs. The `blocked` slice is full of evals from the same job, so the order we pick them off doesn't matter.

Reply:
Well merging a state diagram together with the sequence diagram for the RPCs turned out to be a bad idea. But after a couple hours re-re-reading this code I've got a state diagram here that makes most of the problem clear, which I can annotate.

```mermaid
stateDiagram-v2

    isPending: check if jobEvals\n(pending JobID) is empty
    state if_state <<choice>>

    [*] --> isPending: Enqueue
    isPending --> if_state
    if_state --> ready: no pending JobID.\nready is sorted\nby priority and CreateIndex
    if_state --> blocked: pending JobID

    blocked --> ready: Ack

    ready --> unack: Dequeue
    unack --> [*]: Ack

    unack --> isPending: Nack
```

When we enqueue, we check `jobEvals` (better described as "pending jobs") to see if its non-empty.
- If it's empty, we insert the job ID into `jobEvals` and push the eval onto the `ready` heap.
- If it's non-empty, we insert the eval into `blocked` and return.

When we dequeue, we:
- `scanForSchedulers`:
  - peeks evals from `ready` for each scheduler
  - finds the eval with the highest priortity (or picks between ones with the same)
  - in `dequeueForSched`, pop the selected eval from `ready` and adds it to `unack`
- `getWaitIndex(eval.ns, eval.jobid, eval.ModifyIndex)` for the eval we got:
  - gets highest modify index for any eval for the job, or dequeued eval's modify index, _whichever is greater_

When we submit the plan:
- [updateEvalModifiyInxex](https://github.com/hashicorp/nomad/blob/v1.4.2/nomad/state/state_store.go#L403) updates the eval, but this will not be reflected in eval broker

> // updateEvalModifyIndex is used to update the modify index of an evaluation that has been
> // through a scheduler pass. This is done as part of plan apply. It ensures that when a subsequent
> // scheduler workers process a re-queued evaluation it sees any partial updates from the plan apply.

When we ack the eval:
- remove from `unack`
- delete jobID from `jobEvals` so the job isn't marked "pending"
- enqueue top `blocked` item (which will mark us as "pending" again)

So `ready` sorting is the only thing in the eval broker that cares about `CreateIndex`, and only `getWaitIndex` checks the modify index. As long as we're always setting the eval with the highest modify index as ready, we're good. I _think_ we can cancel the entire `blocked` list when we do so.

```mermaid
stateDiagram-v2

    isPending: check if jobEvals\n(pending JobID) is empty
    state if_state <<choice>>

    [*] --> isPending: Enqueue
    isPending --> if_state
    if_state --> ready: no pending JobID.\nready is sorted\nby priority and CreateIndex
    if_state --> blocked: pending JobID

    blocked --> ready: on Ack\neval w/ highest ModifyIndex
    blocked --> canceled: cancel all the rest

    ready --> unack: Dequeue
    unack --> [*]: Ack

    unack --> isPending: Nack
```

Alternately, if we wanted to be extra paranoid to ensure we haven't missed something we could cancel all but one eval out of `blocked`. That still helps us for the "large number of blocked evals" scenario but makes for bimodal operation where we won't hit this new cancel workflow except in cases where there's a backlog, which I don't love.

Reply:
I'll implement this and make sure we've got an integration test that exercises it.

Reply:
Started on implementation and immediately remembered that evals have their own priority, so we can't just naively use the highest modify index either: we could get a higher-priority eval (ex 100) and then a lower-priority one (ex 50). If we just use highest modify index the lower-priority eval would be the one that gets set to ready and we'd discard the high-priority one. Then that low-priority eval could get starved by middling-priority evals for other jobs (ex 70). We'll need to account for both fields.